src/docfx/build/xref/InternalXrefMap.cs

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal class InternalXrefMap
+    {
+        private readonly IReadOnlyDictionary<string, XrefSpec> _map;
+
+        private InternalXrefMap(IReadOnlyDictionary<string, XrefSpec> map)
+        {
+            _map = map;
+        }
+
+        public static async Task<InternalXrefMap> Create(Context context, IEnumerable<Document> files)
+        {
+            ConcurrentDictionary<string, ConcurrentBag<XrefSpec>> xrefConflicts = new ConcurrentDictionary<string, ConcurrentBag<XrefSpec>>();
+            Debug.Assert(files != null);
+            using (Progress.Start("Building Xref map"))
+            {
+                await ParallelUtility.ForEach(files.Where(f => f.ContentType == ContentType.Page), file => Load(context, xrefConflicts, file), Progress.Update);
+                var xrefs = HandleXrefConflicts(context, xrefConflicts);
+                return new InternalXrefMap(xrefs);
+            }
+        }
+
+        public bool Resolve(string uid, out XrefSpec xref) => _map.TryGetValue(uid, out xref);
+
+        private static Task Load(Context context, ConcurrentDictionary<string, ConcurrentBag<XrefSpec>> xrefConflicts, Document file)
```

Comment:
you can add a `ParallelUtility.ForEach` that takes void return methods

Reply:
Will do this in an additional PR, fixing tocmap as well

---
In reply to: [219370560](https://github.com/dotnet/docfx/pull/3437#discussion_r219370560) [](ancestors = 219370560)